Sources/Core/Tracker/Tracker.swift

```diff
@@ -385,7 +385,63 @@ class Tracker: NSObject {
         emitter.resumeTimer()
         session?.startChecker()
     }
-
+    
+    private func decorateLinkErrorTemplate(_ extendedParameterName: String) -> String {
+        "\(extendedParameterName) has been requested in CrossDeviceParameterConfiguration, but it is not set."
+    }
+    
+    func decorateLink(_ url: URL, extendedParameters: CrossDeviceParameterConfiguration? = nil) -> URL? {
+        var userId: String
+        switch self.session?.userId {
+        case .none:
+            logError(message: "\(url) could not be decorated as session.userId is nil")
+            return nil
+        case .some(let id):
+            userId = id
+        }
+        
+        let extendedParameters = extendedParameters ?? CrossDeviceParameterConfiguration()
+        
+        let sessionId = extendedParameters.sessionId ? self.session?.state?.sessionId ?? "" : ""
+        if (extendedParameters.sessionId && sessionId.isEmpty) {
+            logDebug(message: "\(decorateLinkErrorTemplate("sessionId")) Ensure an event has been tracked to generate a session before calling this method.")
+        }
+        
+        let sourceId = extendedParameters.sourceId ? self.appId : ""
+        
+        let sourcePlatform = extendedParameters.sourcePlatform ? devicePlatformToString(self.devicePlatform) : ""
+        
+        let subjectUserId = extendedParameters.subjectUserId ? self.subject?.userId ?? "" : ""
+        if (extendedParameters.subjectUserId && subjectUserId.isEmpty) {
+            logDebug(message: "\(decorateLinkErrorTemplate("subjectUserId")) Ensure SubjectConfiguration.userId has been set on your tracker.")
+        }
+        
+        let reason = extendedParameters.reason ?? ""
+        
+        let spParameters = [
+            userId,
+            String(Int(Date().timeIntervalSince1970 * 1000)),
+            sessionId,
+            subjectUserId.toBase64(),
+            sourceId.toBase64(),
+            sourcePlatform,
+            reason.toBase64()
+        ].joined(separator: ".").trimmingCharacters(in: ["."])
+        
+        var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+        let spQueryParam = URLQueryItem(name: "_sp", value: spParameters)
+        
+        if let index = components?.queryItems?.firstIndex(where: { $0.name == "_sp" }) {
+            // Replace the old query item with the new one
+            components?.queryItems?[index] = spQueryParam
```

Comment:
Nit: I am a tiny bit worried about modifying the array in place, prefer if we could work with them as immutable arrays. I think we could replace the if-else with something like this:

```
components?.queryItems = (components?.queryItems?.filter { $0.name != "_sp" } ?? []) + [spQueryParam]
```